Client/Application/LeanplumIntegration.swift

```diff
@@ -62,219 +62,178 @@ enum LeanplumEventName: String {
+            }
+        }
     }
 
-    func track(eventName: LeanplumEventName, withParameters parameters: [String: AnyObject]) {
-        DispatchQueue.main.async(execute: {
+    func set(attributes: [AnyHashable: Any]) {
+        DispatchQueue.main.ensureMainThread {
             if self.shouldSendToLP() {
-                Leanplum.track(eventName.rawValue, withParameters: parameters)
+                Leanplum.setUserAttributes(attributes)
             }
-        })
+        }
     }
 
-    // Utils
-    
-    func setEnabled(_ enabled: Bool) {
-        guard AppConstants.MOZ_ENABLE_LEANPLUM else {
-            return
-        }
+    func set(enabled: Bool) {
         // Setting up Test Mode stops sending things to server.
         if enabled { start() }
         self.enabled = enabled
         Leanplum.setTestModeEnabled(!enabled)
     }
 
-    func canInstallFocus() -> Bool {
-        guard let focus = URL(string: "firefox-focus://") else {
-            return false
+    /*
+     We use this to check if an app was installed _after_ a user has installed firefox
+     To do this we only report when the key changes from false -> true
+     If the key is not present we use isAppInstalled bool to set a default (if the app is not installed this will be false)
+     On subsequent launches we check if the users pref is false and if the app is installed (via isAppInstalled)
+     if the value is true the we fire the event!
+     */
+    private func checkIfAppInstalled(key: String, isAppInstalled: Bool, lpEvent: LPEvent) {
+        if self.profile?.prefs.boolForKey(key) == nil {
+            self.profile?.prefs.setBool(isAppInstalled, forKey: key)
         }
-        return !UIApplication.shared.canOpenURL(focus)
-    }
 
-    func canInstallKlar() -> Bool {
-        guard let klar = URL(string: "firefox-klar://") else {
-            return false
+        if !(self.profile?.prefs.boolForKey(key) ?? false), isAppInstalled {
+            self.profile?.prefs.setBool(isAppInstalled, forKey: key)
+            self.track(event: lpEvent)
         }
-        return !UIApplication.shared.canOpenURL(klar)
     }
 
-    func canInstallPocket() -> Bool {
-        guard let pocket = URL(string: "pocket://") else {
-            return false
-        }
-        return !UIApplication.shared.canOpenURL(pocket)
+    private func focusInstalled() -> Bool {
+        return URL(string: "firefox-focus://").flatMap { UIApplication.shared.canOpenURL($0) } ?? false
+    }
+
+    private func klarInstalled() -> Bool {
+        return URL(string: "firefox-klar://").flatMap { UIApplication.shared.canOpenURL($0) } ?? false
+    }
+
+    private func pocketInstalled() -> Bool {
+        return URL(string: "pocket://").flatMap { UIApplication.shared.canOpenURL($0) } ?? false
     }
 
-    func mailtoIsDefault() -> Bool {
+    private func mailtoIsDefault() -> Bool {
         if let option = self.profile?.prefs.stringForKey(PrefsKeys.KeyMailToOption), option != "mailto:" {
```

Comment:
I dont think that would work if the flag has not yet been set. The key would be nil and this would return false.

Reply:
I think a lot of how this code treats nils is implicit and because of how swift lets you simplify nil checking the code looks over verbose. But the nil checking and the order is critical 😞